File: app/src/main/java/arcus/app/subsystems/favorites/adapters/FavoritesListAdapter.java
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.subsystems.favorites.adapters;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.iris.client.model.DeviceModel;
import arcus.app.R;
import arcus.app.common.adapters.AbstractListAdapter;
import arcus.app.common.image.ImageManager;
import arcus.app.common.image.picasso.transformation.BlackWhiteInvertTransformation;
import arcus.app.common.image.picasso.transformation.Invert;


public class FavoritesListAdapter extends AbstractListAdapter<FavoriteItemModel, RecyclerView.ViewHolder> {

    private static final int TYPE_ADD = 0;
    private static final int TYPE_FAVORITE = 1;

    private final Context mContext;

    public FavoritesListAdapter(Context context) {
        mContext = context;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
        switch (viewType){
            case TYPE_FAVORITE:
                return new FavoritesViewHolder(inflater.inflate(R.layout.cell_dashboard_favorites, viewGroup, false));
            default:
                return new AddToFavoritesViewHolder(inflater.inflate(R.layout.add_to_favorites_view, viewGroup, false));
        }
    }

    @Override
    public int getItemViewType(int position) {
        return (position == mData.size() ? TYPE_ADD : TYPE_FAVORITE);
    }

    @Override
    public int getItemCount() {
        return mData.size() + 1;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch(holder.getItemViewType()) {
            case TYPE_FAVORITE:
                ((FavoritesViewHolder) holder).bind(mData.get(position));
                break;
            default:
                ((AddToFavoritesViewHolder) holder).bind();
        }
    }

    public class FavoritesViewHolder extends RecyclerView.ViewHolder {
        @NonNull private final TextView mTextView;
        @NonNull private final ImageView mImageView;

        public FavoritesViewHolder(@NonNull View v) {
            super(v);
            mTextView = (TextView) v.findViewById(R.id.label);
            mImageView = (ImageView) v.findViewById(R.id.image);
        }

        public void bind(@NonNull FavoriteItemModel favoriteItemModel) {
            mTextView.setText(favoriteItemModel.getTitle());
            mTextView.setAlpha(favoriteItemModel.isDisabled() ? 0.4f : 1.0f);

            if (favoriteItemModel.getImageResource() != null) {
                BlackWhiteInvertTransformation transform = new BlackWhiteInvertTransformation(Invert.BLACK_TO_WHITE);
                Bitmap bitmap = BitmapFactory.decodeResource(mContext.getResources(), favoriteItemModel.getImageResource());
                if(favoriteItemModel.getKeepImageColor()) {
                    mImageView.setImageBitmap(bitmap);
                }
                else {
                    mImageView.setImageBitmap(transform.transform(bitmap));
                }
            } else {
                ImageManager.with(mContext)
                        .putSmallDeviceImage((DeviceModel) favoriteItemModel.getModel())
                        .withTransformForStockImages(new BlackWhiteInvertTransformation(Invert.BLACK_TO_WHITE))
                        .withPlaceholder(R.drawable.device_list_placeholder)
                        .withError(R.drawable.device_list_placeholder)
                        .noUserGeneratedImagery()
                        .into(mImageView)
                        .execute();
            }

            mImageView.setAlpha(favoriteItemModel.isDisabled() ? 0.4f : 1.0f);
        }

        @NonNull
        public TextView getTextView() {
            return mTextView;
        }

        @NonNull
        @Override
        public String toString() {
            return "FavoritesViewHolder{" + mTextView.getText() + "}";
        }
    }

    public class AddToFavoritesViewHolder extends RecyclerView.ViewHolder {
        private final AppCompatImageView image;

        AddToFavoritesViewHolder(View itemView) {
            super(itemView);
            image = itemView.findViewById(R.id.image);
        }

        public void bind() {
            image.setVisibility(View.VISIBLE);
            ImageManager.with(itemView.getContext())
                .putDrawableResource(R.drawable.favorite_45x45)
                .withTransformForStockImages(new BlackWhiteInvertTransformation(Invert.BLACK_TO_WHITE))
                .into(image)
                .execute();
         }
    }
}

File: app/src/main/java/arcus/app/dashboard/settings/favorites/FavoritesListAdapter.kt
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.dashboard.settings.favorites

import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.google.common.collect.ImmutableList
import com.h6ah4i.android.widget.advrecyclerview.draggable.DraggableItemAdapter
import com.h6ah4i.android.widget.advrecyclerview.draggable.ItemDraggableRange
import arcus.cornea.subsystem.favorites.FavoritesController
import com.iris.client.model.DeviceModel
import com.iris.client.model.Model
import com.iris.client.model.SceneModel
import arcus.app.R
import arcus.app.common.image.ImageManager
import arcus.app.common.image.picasso.transformation.BlackWhiteInvertTransformation
import arcus.app.common.image.picasso.transformation.Invert
import arcus.app.common.utils.GlobalSetting
import arcus.app.common.utils.ImageUtils
import arcus.app.common.utils.PreferenceUtils
import arcus.app.common.utils.ViewUtils
import arcus.app.subsystems.scenes.catalog.model.SceneCategory
import de.greenrobot.event.EventBus

class FavoritesListAdapter(private val context: Context, private val provider: FavoritesListDataProvider)
    : RecyclerView.Adapter<FavoritesListItemViewHolder>(),
        DraggableItemAdapter<FavoritesListItemViewHolder>,
        FavoritesListDataProvider.OnDataChanged {

    val orderedDeviceIdList: List<String>
        get() = provider.orderedDeviceIdList

    init {
        this.provider.setCallback(this)
        setHasStableIds(true)
    }

    override fun getItemId(position: Int): Long {
        return provider.getItem(position).id
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FavoritesListItemViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val v = inflater.inflate(R.layout.card_favorites_list_item, parent, false)
        return FavoritesListItemViewHolder(v)
    }

    override fun onBindViewHolder(holder: FavoritesListItemViewHolder, position: Int) {
        val item = provider.getItem(position)

        holder.deviceTitle.text = item.modelName

        when (FavoritesController.determineModelType(item.model.address)) {
            FavoritesController.DEVICE_MODEL -> ImageManager.with(context)
                .putSmallDeviceImage(item.model as DeviceModel)
                .noUserGeneratedImagery()
                .withTransformForStockImages(BlackWhiteInvertTransformation(Invert.WHITE_TO_BLACK))
                .into(holder.deviceIcon)
                .execute()
            FavoritesController.SCENE_MODEL -> ImageManager.with(context)
                .putDrawableResource(SceneCategory.fromSceneModel(item.model as SceneModel).iconResId)
                .withTransformForStockImages(BlackWhiteInvertTransformation(Invert.WHITE_TO_BLACK))
                .into(holder.deviceIcon)
                .execute()
        }

        holder.favoriteIcon.setImageResource(
            when {
                deviceIsFavorite(item.model) -> R.drawable.favorite_teal_fill_22x20
                else -> R.drawable.favorite_teal_outline_22x20
            }
        )

        holder.favoriteIcon.setOnClickListener {
            val deviceModel = item.model
            val favoriteTag = ImmutableList.of(GlobalSetting.FAVORITE_TAG)

            if (deviceIsFavorite(deviceModel)) {
                deviceModel.removeTags(favoriteTag)
                provider.removeItem(holder.adapterPosition)
                notifyDataSetChanged()
            } else {
                deviceModel.addTags(favoriteTag)
            }

            deviceModel.commit()
        }

        // Adjust the favorites icon hitbox to margins
        val horizontalMargin = 22
        val verticalMargin = (holder.deviceIcon.height - holder.favoriteIcon.height / 2) + 10

        ViewUtils.increaseTouchArea(holder.favoriteIcon, -horizontalMargin, -verticalMargin, horizontalMargin, verticalMargin)
    }

    /**
     * Logs
     * dashboard.settings.favorites.edit.reorder	User reordered favorite order on dashboard -> settings -> favorites -> edit screen
     */
    override fun onMoveItem(fromPosition: Int, toPosition: Int) {
        if (fromPosition == toPosition) { return }

        provider.moveItem(fromPosition, toPosition)
        notifyItemMoved(fromPosition, toPosition)
    }

    override fun updated() { notifyDataSetChanged() }

    override fun onItemDragFinished(fromPosition: Int, toPosition: Int, result: Boolean) {
        EventBus.getDefault().postSticky(FavoritesOrderChangedEvent(orderedDeviceIdList))
        PreferenceUtils.putOrderedFavoritesList(orderedDeviceIdList)
    }

    private fun deviceIsFavorite(model: Model): Boolean { return model.tags.contains(GlobalSetting.FAVORITE_TAG) }

    override fun getItemCount(): Int { return provider.count }

    override fun onCheckCanStartDrag(holder: FavoritesListItemViewHolder, position: Int, x: Int, y: Int): Boolean {
        return ImageUtils.isLeftOfView(x, holder.deviceIcon, 0)
    }

    override fun onCheckCanDrop(draggingPosition: Int, dropPosition: Int): Boolean { return false }

    override fun onGetItemDraggableRange(favoritesListItemViewHolder: FavoritesListItemViewHolder, i: Int): ItemDraggableRange? { return null }

    override fun onItemDragStarted(position: Int) { /* nop */ }
}
